File: product-service/src/main/java/com/heulwen/productservice/service/CategoryService.java
package com.heulwen.productservice.service;

import com.heulwen.productservice.api.request.CategoryRequest;
import com.heulwen.productservice.api.response.CategoryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface CategoryService {
    CategoryResponse createCategory(CategoryRequest request);
    CategoryResponse updateCategory(Long id, CategoryRequest request);
    void deleteCategory(Long id);
    CategoryResponse getCategory(Long id);
    Page<CategoryResponse> getCategories(String search, Pageable pageable);
}

File: frontend-service/src/services/categoryService.ts
import { apiFetch } from './api';
import { Category, CategoryRequest } from '../types/category';
import { PageResponse } from '../types/common';

export const categoryService = {
  async getCategories(search?: string, page = 0, size = 10): Promise<PageResponse<Category>> {
    return apiFetch<PageResponse<Category>>('/api/categories', {
      method: 'GET',
      params: { search, page, size },
    });
  },

  async getCategory(id: number): Promise<Category> {
    return apiFetch<Category>(`/api/categories/${id}`, {
      method: 'GET',
    });
  },

  async createCategory(request: CategoryRequest): Promise<Category> {
    return apiFetch<Category>('/api/categories', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  },

  async updateCategory(id: number, request: CategoryRequest): Promise<Category> {
    return apiFetch<Category>(`/api/categories/${id}`, {
      method: 'PUT',
      body: JSON.stringify(request),
    });
  },

  async deleteCategory(id: number): Promise<void> {
    return apiFetch<void>(`/api/categories/${id}`, {
      method: 'DELETE',
    });
  }
};
